transport/tlsfrag/tls.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsfrag
+
+import (
+	"encoding/binary"
+)
+
+// This file contains helper functions and constants for TLS Client Hello message.
+
+type recordType byte
+type tlsVersion uint16
+type tlsRecordHeader []byte
+
+// TLS record layout from [RFC 8446]:
+//
+//	+-------------+ 0
+//	| RecordType  |
+//	+-------------+ 1
+//	|  Protocol   |
+//	|  Version    |
+//	+-------------+ 3
+//	|   Record    |
+//	|   Length    |
+//	+-------------+ 5
+//	|   Message   |
+//	|    Data     |
+//	|     ...     |
+//	+-------------+ Message Length + 5
+//
+//	RecordType := invalid(0) | handshake(22) | application_data(23) | ...
+//	LegacyRecordVersion := 0x0301 ("TLS 1.0") | 0x0302 ("TLS 1.1") | 0x0303 ("TLS 1.2")
+//	0 < Message Length (of handshake)        ≤ 2^14
+//	0 ≤ Message Length (of application_data) ≤ 2^14
+//
+// [RFC 8446]: https://datatracker.ietf.org/doc/html/rfc8446#section-5.1
+const (
+	recordHeaderLen     = 5
+	maxRecordPayloadLen = 1 << 14
+
+	recordTypeHandshake recordType = 22
+
+	versionTLS10 tlsVersion = 0x0301
+	versionTLS11 tlsVersion = 0x0302
+	versionTLS12 tlsVersion = 0x0303
+	versionTLS13 tlsVersion = 0x0304
+)
+
+// tlsRecordHeaderFromRawBytes creates a new tlsRecordHeader from raw. This function will panic if len(raw) < 5.
+func tlsRecordHeaderFromRawBytes(raw []byte) tlsRecordHeader {
```

Comment:
I've tried but did not figure out a simpler way of handling updates. Because `[5]byte` is an array, so it will be value-copied, for example after `g := tlsRecordHeader(b.data)`, `g` is a copy of `b.data` instead of a reference, so any updates to `g` won't be reflected to `b.data`, we have to introduced another helper method `func (h tlsRecordHeader) PutToSlice(data []byte)` to copy all bytes from `h` to `data`, that's more complicated if we simply define `tlsRecordHeader` as a slice.